File: latexdraw-core/net.sf.latexdraw/src/main/net/sf/latexdraw/instruments/Pencil.scala
package net.sf.latexdraw.instruments

import java.awt.Component
import java.awt.Point
import java.util.List
import java.util.Objects

import javax.swing.JFileChooser
import javax.swing.SwingUtilities

import net.sf.latexdraw.actions.shape.AddShape
import net.sf.latexdraw.actions.shape.InitTextSetter
import net.sf.latexdraw.actions.shape.InsertPicture
import net.sf.latexdraw.badaboom.BadaboomCollector
import net.sf.latexdraw.filters.PictureFilter
import net.sf.latexdraw.glib.models.ShapeFactory;
import net.sf.latexdraw.glib.models.interfaces.shape._
import net.sf.latexdraw.glib.models.interfaces.shape.IShape.BorderPos
import net.sf.latexdraw.glib.ui.ICanvas
import net.sf.latexdraw.glib.ui.LMagneticGrid
import net.sf.latexdraw.glib.views.Java2D.interfaces.IViewShape
import net.sf.latexdraw.glib.views.Java2D.interfaces.View2DTK
import net.sf.latexdraw.util.LNumber

import org.malai.instrument.Instrument
import org.malai.instrument.Link
import org.malai.interaction.Interaction
import org.malai.interaction.library.Press
import org.malai.stateMachine.MustAbortStateMachineException
import org.malai.swing.instrument.library.WidgetZoomer
import org.malai.swing.interaction.library.AbortableDnD
import org.malai.swing.interaction.library.MultiClick
import org.malai.swing.widget.MLayeredPane

/**
 * This instrument allows to draw shapes.<br>
 * <br>
 * This file is part of LaTeXDraw<br>
 * Copyright (c) 2005-2014 Arnaud BLOUIN<br>
 * <br>
 *  LaTeXDraw is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.<br>
 * <br>
 *  LaTeXDraw is distributed without any warranty; without even the
 *  implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
 *  PURPOSE. See the GNU General Public License for more details.<br>
 * <br>
 * 05/13/10<br>
 * @author Arnaud BLOUIN
 * @version 3.0
 */
class Pencil(canvas : ICanvas, val textSetter:TextSetter, val layers:MLayeredPane) extends CanvasInstrument(canvas) {
	/** The current editing choice (rectangle, ellipse, etc.) of the instrument. */
	var _currentChoice = EditionChoice.RECT

	/** The file chooser used to select pictures. */
	protected lazy val _pictureFileChooser = {
		val fc = new JFileChooser()
		fc.setMultiSelectionEnabled(false)
		fc.setAcceptAllFileFilterUsed(true)
		fc.setFileFilter(new PictureFilter)
		fc
	}

	protected lazy val _groupParams = {
		val group = ShapeFactory.createGroup(false)
		group.addShape(ShapeFactory.createRectangle(false))
		group.addShape(ShapeFactory.createDot(ShapeFactory.createPoint, false))
		group.addShape(ShapeFactory.createGrid(false, ShapeFactory.createPoint))
		group.addShape(ShapeFactory.createAxes(false, ShapeFactory.createPoint))
		group.addShape(ShapeFactory.createText(false))
		group.addShape(ShapeFactory.createCircleArc(false))
		group.addShape(ShapeFactory.createPolyline(false))
		group.addShape(ShapeFactory.createBezierCurve(false))
		group.addShape(ShapeFactory.createFreeHand(false))
		group
	}

	override def setActivated(activated:Boolean) {
		if(this.activated!=activated)
			super.setActivated(activated)
	}

	override def interimFeedback() {
		canvas.setTempView(null)
		canvas.refresh
	}

	override protected def initialiseLinks() {
		try{
			addLink(new Press2AddShape(this))
			addLink(new Press2AddText(this))
			addLink(new Press2InsertPicture(this))
			addLink(new DnD2AddShape(this))
			addLink(new MultiClic2AddShape(this))
			addLink(new Press2InitTextSetter(this))
		}catch{ case e : Exception => BadaboomCollector.INSTANCE.add(e) }
	}

	/**
	 * @return The shape that contains the parameters of the pencil. These parameters
	 * will be used for the creation of new shapes.
	 * @since 3.0
	 */
	def groupParams() = _groupParams

	/**
	 * @return An instance of a shape configured (thickness, colours, etc.) with the parameters of the pencil.
	 * @since 3.0
	 */
	def createShapeInstance() = setShapeParameters(_currentChoice.createShapeInstance)

	/**
	 * Configures the given shape with the parameters (e.g. thickness, colours, etc.) of the pencil.
	 * @param shape The shape to configure.
	 * @return The modified shape given as argument.
	 * @since 3.0
	 */
	def setShapeParameters(shape:IShape) = {
		if(shape.isInstanceOf[IModifiablePointsShape] && !shape.isInstanceOf[IFreehand]) {//FIXME
			val mod = shape.asInstanceOf[IModifiablePointsShape]
			mod.addPoint(ShapeFactory.createPoint)
			mod.addPoint(ShapeFactory.createPoint)
		}

		shape.copy(groupParams)
		shape.setModified(true)
		shape
	}

	/** @return The file chooser used to select pictures. */
	def pictureFileChooser() = _pictureFileChooser

	/** @return The current editing choice. */
	def currentChoice() = _currentChoice

	/**
	 * Sets the current editing choice.
	 * @param currentChoice The new editing choice to set.
	 * @since 3.0
	 */
	def setCurrentChoice(currentChoice:EditionChoice) {
		_currentChoice = currentChoice
	}
}



/**
 * This class defines a generic link for the pencil.
 * @author Arnaud Blouin
 * @since 3.0
 * @version 3.0
 */
private abstract sealed class PencilLink[I <: Interaction](pencil:Pencil, exec:Boolean, clazzInteraction:Class[I])
				extends Link[AddShape, I, Pencil](pencil, false, classOf[AddShape], clazzInteraction) {
	protected var tmpShape : IViewShape = _

	override def initAction() {
		val sh = instrument.createShapeInstance
		tmpShape = View2DTK.getFactory.createView(sh)
		action.setShape(sh)
		action.setDrawing(instrument.canvas.getDrawing)
		instrument.canvas.setTempView(tmpShape)
	}

	override def interimFeedback() {
		tmpShape.update
		instrument.canvas.refresh
	}
}



/**
 * This link allows to create shapes using a multi-clic interaction.
 */
private sealed class MultiClic2AddShape(pencil:Pencil) extends PencilLink[MultiClick](pencil, false, classOf[MultiClick]) {
	// To avoid the overlapping with the DnD2AddShape, the starting interaction must be
	// aborted when the condition is not respected, i.e. when the selected shape type is
	// not devoted to the multi-clic interaction.
	override def isInteractionMustBeAborted() = !isConditionRespected

	override def updateAction() {
		val pts	= interaction.getPoints
		val currPoint = instrument.getAdaptedPoint(interaction.getCurrentPosition)
		val shape = action.shape.get.asInstanceOf[IModifiablePointsShape]

		if(shape.getNbPoints==pts.size && !interaction.isLastPointFinalPoint)
			shape.addPoint(ShapeFactory.createPoint(currPoint.getX, currPoint.getY), pts.size-1)
		else
			shape.setPoint(currPoint.getX, currPoint.getY, -1)

		// Curves need to be balanced.
		if(shape.isInstanceOf[IControlPointShape])
			shape.asInstanceOf[IControlPointShape].balance

		shape.setModified(true)
		action.drawing.get.setModified(true)
	}


	override def initAction() {
		super.initAction
		val shape = action.shape.get

		if(shape.isInstanceOf[IModifiablePointsShape]) {
			val modShape = shape.asInstanceOf[IModifiablePointsShape]
			val pt = instrument.getAdaptedPoint(interaction.getPoints.get(0))
			modShape.setPoint(pt, 0)
			modShape.setPoint(pt.getX+1, pt.getY+1, 1)
		}
	}

	override def isConditionRespected() = instrument.currentChoice==EditionChoice.POLYGON ||
			instrument.currentChoice==EditionChoice.LINES || instrument.currentChoice==EditionChoice.BEZIER_CURVE ||
			instrument.currentChoice==EditionChoice.BEZIER_CURVE_CLOSED

	override def interactionStarts(inter:Interaction) {
		super.interactionStarts(inter)
		if(instrument.currentChoice==EditionChoice.POLYGON)
			interaction.setMinPoints(3)
		else
			interaction.setMinPoints(2)
	}
}



/**
 * This link allows to create shapes using a drag-and-drop interaction.
 */
private sealed class DnD2AddShape(pencil:Pencil) extends PencilLink[AbortableDnD](pencil, false, classOf[AbortableDnD]) {
	override def initAction() {
		super.initAction
		action.shape match {
			case Some(shape) =>
				val ec = instrument.currentChoice
				val pt = instrument.getAdaptedPoint(interaction.getStartPt)

				// For squares and circles, the centre of the shape is the reference point during the creation.
				shape match {
					case sq:ISquaredShape =>
							sq.setPosition(pt.getX-1, pt.getY-1)
							sq.setWidth(2.0)
					case fh:IFreehand => fh.addPoint(ShapeFactory.createPoint(pt.getX, pt.getY))
					case _ => shape.translate(pt.getX, pt.getY)
				}
			case _ =>
		}
	}


	override def isConditionRespected() = {
		val ec = instrument.currentChoice
		ec==EditionChoice.RECT || ec==EditionChoice.ELLIPSE || ec==EditionChoice.SQUARE || ec==EditionChoice.CIRCLE ||
		ec==EditionChoice.RHOMBUS || ec==EditionChoice.TRIANGLE  || ec==EditionChoice.CIRCLE_ARC || ec==EditionChoice.FREE_HAND;
	}


	override def updateAction() {
		val shape = action.shape.get
		val ec = instrument.currentChoice
		// Getting the points depending on the current zoom.
		val startPt = instrument.getAdaptedPoint(interaction.getStartPt)
		val endPt   = instrument.getAdaptedPoint(interaction.getEndPt)

		shape match {
			case sq:ISquaredShape =>
					updateShapeFromCentre(sq, startPt, endPt.getX)
					shape.setModified(true)
					action.drawing.get.setModified(true)
			case fh:IFreehand =>
				val lastPoint = fh.getPtAt(-1)
				if(!LNumber.equalsDouble(lastPoint.getX, endPt.getX, 0.0001) && !LNumber.equalsDouble(lastPoint.getY, endPt.getY, 0.0001))
					fh.addPoint(endPt)
			case rec:IRectangularShape =>
					updateShapeFromDiag(rec, startPt, endPt)
					shape.setModified(true)
					action.drawing.get.setModified(true)
			case _ =>
		}
	}


	/**
	 * @param shape The shape to analyse.
	 * @return The gap that must respect the pencil to not allow shape to disappear when they are too small.
	 * @since 3.0
	 */
	private def getGap(shape:IShape) : Double = {
		// These lines are necessary to avoid shape to disappear. It appends when the borders position is INTO. In this case, the
		// minimum radius must be computed using the thickness and the double size.
		if(shape.isBordersMovable && shape.getBordersPosition()==BorderPos.INTO)
			shape.getThickness + (if(shape.isDbleBorderable && shape.hasDbleBord) shape.getDbleBordSep else 0.0)
		else 1.0
	}


	private def updateShapeFromCentre(shape:ISquaredShape, startPt:IPoint, endX:Double) {
		val sx = startPt.getX
		val radius = Math.max(if(sx<endX) endX-sx else sx - endX, getGap(shape))
		shape.setPosition(sx-radius, startPt.getY+radius)
		shape.setWidth(radius*2.0)
	}


	private def updateShapeFromDiag(shape:IRectangularShape, startPt:IPoint, endPt:IPoint) {
		val gap = getGap(shape)
		var v1 = startPt.getX
		var v2 = endPt.getX
		var tlx = Double.NaN
		var tly = Double.NaN
		var brx = Double.NaN
		var bry = Double.NaN
		var ok = true

		if(Math.abs(v1-v2)>gap)
			if(v1<v2) {
				brx = v2
				tlx = v1
			}
			else {
				brx = v1
				tlx = v2
			}
		else ok = false

		v1 = startPt.getY
		v2 = endPt.getY

		if(Math.abs(v1-v2)>gap)
			if(v1<v2) {
				bry = v2
				tly = v1
			}
			else {
				bry = v1
				tly = v2
			}
		else ok = false

		if(ok) {
			shape.setPosition(tlx, bry)
			shape.setWidth(brx - tlx)
			shape.setHeight(bry - tly)
		}
	}
}



/**
 * Maps a mouse press interaction to an action that asks and adds a picture into the drawing.
 */
private sealed class Press2InsertPicture(pencil:Pencil) extends Link[InsertPicture, Press, Pencil](pencil, false, classOf[InsertPicture], classOf[Press]) {
	override def initAction() {
		action.setDrawing(instrument.canvas.getDrawing)
		action.setShape(ShapeFactory.createPicture(true, instrument.getAdaptedPoint(interaction.getPoint)))
		action.setFileChooser(instrument.pictureFileChooser)
	}

	override def isConditionRespected() = instrument.currentChoice==EditionChoice.PICTURE
}



private sealed class Press2AddShape(pencil:Pencil) extends PencilLink[Press](pencil, false, classOf[Press]) {
	override def initAction() {
		super.initAction
		val shape = action.shape.get
		if(shape.isInstanceOf[IPositionShape]) {
			shape.asInstanceOf[IPositionShape].setPosition(instrument.getAdaptedPoint(interaction.getPoint))
			shape.setModified(true)
		}
	}

	override def isConditionRespected() = instrument.currentChoice==EditionChoice.GRID || instrument.currentChoice==EditionChoice.DOT || instrument.currentChoice==EditionChoice.AXES;
}



/**
 * When a user starts to type a text using the text setter and then he clicks somewhere else
 * in the canvas, the text typed must be added (if possible to the canvas) before starting typing
 * a new text.
 */
private sealed class Press2AddText(pencil:Pencil) extends PencilLink[Press](pencil, false, classOf[Press]) {
	override def initAction() {
		action.setDrawing(instrument.canvas.getDrawing)
		action.setShape(ShapeFactory.createText(true, ShapeFactory.createPoint(instrument.textSetter.relativePoint), instrument.textSetter.getTextField.getText))
	}

	// The action is created only when the user uses the text setter and the text field of the text setter is not empty.
	override def isConditionRespected() = instrument.textSetter.isActivated && instrument.textSetter.getTextField.getText.length>0
}


/**
 * This link maps a press interaction to activate the instrument
 * that allows to add and modify some texts.
 */
private sealed class Press2InitTextSetter(pencil:Pencil) extends Link[InitTextSetter, Press, Pencil](pencil, false, classOf[InitTextSetter], classOf[Press]) {
	override def initAction() {
		action.setText("")
		action.setTextShape(null)
		action.setInstrument(instrument.textSetter)
		action.setTextSetter(instrument.textSetter)
		action.setAbsolutePoint(ShapeFactory.createPoint(
				SwingUtilities.convertPoint(interaction.getTarget.asInstanceOf[Component],interaction.getPoint, instrument.layers)))
		action.setRelativePoint(instrument.getAdaptedPoint(interaction.getPoint))
	}

	override def isConditionRespected() = instrument.currentChoice==EditionChoice.TEXT
}

File: latexdraw-core/net.sf.latexdraw/src/main/java/net/sf/latexdraw/instruments/Pencil.java
/*
 * This file is part of LaTeXDraw.
 * Copyright (c) 2005-2018 Arnaud BLOUIN
 * LaTeXDraw is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 2 of the License, or (at your option) any later version.
 * LaTeXDraw is distributed without any warranty; without even the implied
 * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 */
package net.sf.latexdraw.instruments;

import java.util.Arrays;
import java.util.Collections;
import java.util.function.Function;
import javafx.application.Platform;
import javafx.beans.property.ObjectProperty;
import javafx.beans.property.SimpleObjectProperty;
import javafx.geometry.Point3D;
import javafx.scene.Cursor;
import javafx.scene.input.MouseButton;
import javafx.stage.FileChooser;
import net.sf.latexdraw.commands.shape.AddShape;
import net.sf.latexdraw.commands.shape.InitTextSetter;
import net.sf.latexdraw.commands.shape.InsertPicture;
import net.sf.latexdraw.models.MathUtils;
import net.sf.latexdraw.models.ShapeFactory;
import net.sf.latexdraw.models.interfaces.shape.BorderPos;
import net.sf.latexdraw.models.interfaces.shape.IBezierCurve;
import net.sf.latexdraw.models.interfaces.shape.IControlPointShape;
import net.sf.latexdraw.models.interfaces.shape.IFreehand;
import net.sf.latexdraw.models.interfaces.shape.IGroup;
import net.sf.latexdraw.models.interfaces.shape.IModifiablePointsShape;
import net.sf.latexdraw.models.interfaces.shape.IPoint;
import net.sf.latexdraw.models.interfaces.shape.IPolygon;
import net.sf.latexdraw.models.interfaces.shape.IPolyline;
import net.sf.latexdraw.models.interfaces.shape.IPositionShape;
import net.sf.latexdraw.models.interfaces.shape.IRectangularShape;
import net.sf.latexdraw.models.interfaces.shape.IShape;
import net.sf.latexdraw.models.interfaces.shape.ISquaredShape;
import net.sf.latexdraw.util.Inject;
import net.sf.latexdraw.view.jfx.ViewFactory;
import org.malai.javafx.interaction.library.DnD;
import org.malai.javafx.interaction.library.MultiClick;
import org.malai.javafx.interaction.library.PointsData;
import org.malai.javafx.interaction.library.Press;

/**
 * This instrument allows to draw shapes.
 * @author Arnaud Blouin
 */
public class Pencil extends CanvasInstrument {
	/** The current editing choice (rectangle, ellipse, etc.) of the instrument. */
	private final ObjectProperty<EditionChoice> currentChoice;

	/** The text setter used to create text shapes. */
	@Inject protected TextSetter textSetter;

	/** The file chooser used to select the picture to insert. Use its getter instead as it is lazy instantiated. */
	private FileChooser pictureFileChooser;

	/** This shape gathers all the current shape parameters. Used as a model when creating shapes. Use its getter instead as it is lazy instantiated. */
	private IGroup groupParams;


	/**
	 * Creates the pencil. Automatically invoked by the dependency injector. Do not call.
	 */
	public Pencil() {
		super();
		currentChoice = new SimpleObjectProperty<>(EditionChoice.RECT);
	}

	@Override
	public void uninstallBindings() {
		currentChoice.unbind();
		pictureFileChooser = null;
		groupParams = null;
		super.uninstallBindings();
	}

	/**
	 * Entry point for a testing purpose only.
	 */
	protected void setPictureFileChooser(final FileChooser chooser) {
		pictureFileChooser = chooser;
	}

	/**
	 * @return The file chooser used to select pictures to insert. Cannot be null.
	 */
	protected FileChooser getPictureFileChooser() {
		if(pictureFileChooser == null) {
			pictureFileChooser = new FileChooser();
			pictureFileChooser.getExtensionFilters().add(new FileChooser.ExtensionFilter(
				"png/bmp/gif/jpg/eps/pdf", Arrays.asList("*.png", "*.bmp", "*.gif", "*.jpeg", "*.jpg", "*.eps", "*.pdf"))); //NON-NLS
		}
		return pictureFileChooser;
	}

	/**
	 * @return the shape that gathers all the current shape parameters. Use as a model to copy while creating a new shape. Cannot be null.
	 */
	public IGroup getGroupParams() {
		if(groupParams == null) {
			groupParams = ShapeFactory.INST.createGroup();
			groupParams.addShape(ShapeFactory.INST.createRectangle());
			groupParams.addShape(ShapeFactory.INST.createDot(ShapeFactory.INST.createPoint()));
			groupParams.addShape(ShapeFactory.INST.createGrid(ShapeFactory.INST.createPoint()));
			groupParams.addShape(ShapeFactory.INST.createAxes(ShapeFactory.INST.createPoint()));
			groupParams.addShape(ShapeFactory.INST.createText());
			groupParams.addShape(ShapeFactory.INST.createCircleArc());
			groupParams.addShape(ShapeFactory.INST.createPolyline(Collections.emptyList()));
			groupParams.addShape(ShapeFactory.INST.createBezierCurve(Collections.emptyList()));
			groupParams.addShape(ShapeFactory.INST.createFreeHand(Collections.emptyList()));
			groupParams.addShape(ShapeFactory.INST.createPlot(ShapeFactory.INST.createPoint(), 1, 10, "x", false)); //NON-NLS
		}
		return groupParams;
	}

	@Override
	public void setActivated(final boolean active) {
		if(activated != active) {
			super.setActivated(active);
		}
	}

	@Override
	public void interimFeedback() {
		canvas.setTempView(null);
		canvas.setCursor(Cursor.DEFAULT);
	}

	@Override
	protected void configureBindings() {
		bindPressToAddShape();

		// Binds a pressure to insert a picture
		nodeBinder(new Press(), InsertPicture.class).on(canvas).
			map(i -> new InsertPicture(ShapeFactory.INST.createPicture(getAdaptedPoint(i.getSrcLocalPoint())), canvas.getDrawing(), getPictureFileChooser())).
			when(i -> currentChoice.get() == EditionChoice.PICTURE && i.getButton() == MouseButton.PRIMARY).bind();

		bindDnDToDrawRectangularShape();
		bindDnDToDrawSquaredShape();
		bindDnDToDrawFreeHandShape();
		bindMultiClic2AddShape();

		// Binds a pressure to show the text setter
		nodeBinder(new Press(), InitTextSetter.class).on(canvas).
			map(i -> new InitTextSetter(textSetter, textSetter, "", getAdaptedPoint(i.getSrcLocalPoint()), null, null)).
			when(i -> (currentChoice.get() == EditionChoice.TEXT || currentChoice.get() == EditionChoice.PLOT) && i.getButton() == MouseButton.PRIMARY).bind();
	}

	/**
	 * Binds a DnD interaction to create shape.
	 */
	private void bindDnDToDrawFreeHandShape() {
		nodeBinder(new DnD(false, true), AddShape.class).on(canvas).
			map(i -> {
				final IShape sh = createShapeInstance();
				final IPoint pt = getAdaptedPoint(i.getSrcLocalPoint());
				sh.getPoints().get(0).setPoint(pt.getX(), pt.getY());
				return new AddShape(sh, canvas.getDrawing());
			}).
			first((i, c) -> Platform.runLater(() -> canvas.requestFocus())).
			then((i, c) -> {
				final IPoint last = c.getShape().get().getPtAt(-1);
				final IPoint endPt = getAdaptedPoint(i.getTgtLocalPoint());
				if(!MathUtils.INST.equalsDouble(last.getX(), endPt.getX(), 0.0001) &&
					!MathUtils.INST.equalsDouble(last.getY(), endPt.getY(), 0.0001)) {
					c.setShape(ShapeFactory.INST.createFreeHandFrom((IFreehand) c.getShape().get(), endPt));
				}
				canvas.setTempView(ViewFactory.INSTANCE.createView(c.getShape().orElse(null)).orElse(null));
			}).
			endOrCancel((i, c) -> canvas.setTempView(null)).
			when(i -> i.getButton() == MouseButton.PRIMARY && currentChoice.get() == EditionChoice.FREE_HAND).
			strictStart().
			bind();
	}

	/**
	 * Binds a DnD interaction to draw squared shapes.
	 */
	private void bindDnDToDrawSquaredShape() {
		nodeBinder(new DnD(false, true), AddShape.class).on(canvas).
			map(i -> {
				final ISquaredShape sq = (ISquaredShape) createShapeInstance();
				final IPoint pt = getAdaptedPoint(i.getSrcLocalPoint());
				sq.setPosition(pt.getX() - 1d, pt.getY() - 1d);
				sq.setWidth(2d);
				return new AddShape(sq, canvas.getDrawing());
			}).
			first((i, c) -> {
				Platform.runLater(() -> canvas.requestFocus());
				canvas.setTempView(ViewFactory.INSTANCE.createView(c.getShape().orElse(null)).orElse(null));
			}).
			then((i, c) -> updateShapeFromCentre((ISquaredShape) c.getShape().get(), getAdaptedPoint(i.getSrcLocalPoint()), getAdaptedPoint(i.getTgtLocalPoint()).getX())).
			endOrCancel((i, c) -> canvas.setTempView(null)).
			when(i -> i.getButton() == MouseButton.PRIMARY).
			strictStart().
			bind().activationProperty().bind(activatedProp.and(currentChoice.isEqualTo(EditionChoice.SQUARE).or(currentChoice.isEqualTo(EditionChoice.CIRCLE).
				or(currentChoice.isEqualTo(EditionChoice.CIRCLE_ARC)))));
	}

	/**
	 * Binds a DnD interaction to draw rectangular shapes.
	 */
	private void bindDnDToDrawRectangularShape() {
		nodeBinder(new DnD(false, true), AddShape.class).on(canvas).
			map(i -> new AddShape(createShapeInstance(), canvas.getDrawing())).
			first((i, c) -> {
				Platform.runLater(() -> canvas.requestFocus());
				canvas.setTempView(ViewFactory.INSTANCE.createView(c.getShape().orElse(null)).orElse(null));
			}).
			then((i, c) -> updateShapeFromDiag((IRectangularShape) c.getShape().get(), getAdaptedPoint(i.getSrcLocalPoint()), getAdaptedPoint(i.getTgtLocalPoint()))).
			endOrCancel((a, i) -> canvas.setTempView(null)).
			when(i -> i.getButton() == MouseButton.PRIMARY).
			strictStart().
			bind().activationProperty().bind(activatedProp.and(currentChoice.isEqualTo(EditionChoice.RECT).or(currentChoice.isEqualTo(EditionChoice.ELLIPSE).
				or(currentChoice.isEqualTo(EditionChoice.RHOMBUS).or(currentChoice.isEqualTo(EditionChoice.TRIANGLE))))));
	}

	/**
	 * Binds a multi-click interaction to creates multi-point shapes.
	 */
	private void bindMultiClic2AddShape() {
		final Function<PointsData, AddShape> creation = i -> new AddShape(setInitialPtsShape(createShapeInstance(),
			i.getPointsData().get(0).getSrcLocalPoint()), canvas.getDrawing());

		// Binding for polygons
		nodeBinder(new MultiClick(3), AddShape.class).on(canvas).map(creation).
			then((i, c) -> {
				final IPoint currPoint = getAdaptedPoint(i.getCurrentPosition());
				if(c.getShape().get().getNbPoints() == i.getPointsData().size() && i.getLastButton().orElse(MouseButton.NONE) == MouseButton.PRIMARY) {
					c.setShape(ShapeFactory.INST.createPolygonFrom((IPolygon) c.getShape().get(), ShapeFactory.INST.createPoint(currPoint.getX(), currPoint.getY())));
				}else {
					((IModifiablePointsShape) c.getShape().get()).setPoint(currPoint.getX(), currPoint.getY(), -1);
				}
				canvas.setTempView(ViewFactory.INSTANCE.createView(c.getShape().orElse(null)).orElse(null));
			}).
			endOrCancel((i, c) -> canvas.setTempView(null)).
			bind().activationProperty().bind(currentChoice.isEqualTo(EditionChoice.POLYGON).and(activatedProp));

		// Binding for polyline
		nodeBinder(new MultiClick(), AddShape.class).on(canvas).map(creation).
			then((i, c) -> {
				final IPoint currPoint = getAdaptedPoint(i.getCurrentPosition());
				if(c.getShape().get().getNbPoints() == i.getPointsData().size() && i.getLastButton().orElse(MouseButton.NONE) == MouseButton.PRIMARY) {
					c.setShape(ShapeFactory.INST.createPolylineFrom((IPolyline) c.getShape().get(), ShapeFactory.INST.createPoint(currPoint.getX(), currPoint.getY())));
				}else {
					((IModifiablePointsShape) c.getShape().get()).setPoint(currPoint.getX(), currPoint.getY(), -1);
				}
				canvas.setTempView(ViewFactory.INSTANCE.createView(c.getShape().orElse(null)).orElse(null));
			}).
			endOrCancel((i, c) -> canvas.setTempView(null)).
			bind().activationProperty().bind(currentChoice.isEqualTo(EditionChoice.LINES).and(activatedProp));

		// Binding for bézier curves
		nodeBinder(new MultiClick(), AddShape.class).on(canvas).map(creation).
			then((i, c) -> {
				final IPoint currPoint = getAdaptedPoint(i.getCurrentPosition());
				if(c.getShape().get().getNbPoints() == i.getPointsData().size() && i.getLastButton().orElse(MouseButton.NONE) == MouseButton.PRIMARY) {
					c.setShape(ShapeFactory.INST.createBezierCurveFrom((IBezierCurve) c.getShape().get(), ShapeFactory.INST.createPoint(currPoint.getX(), currPoint.getY())));
				}else {
					((IModifiablePointsShape) c.getShape().get()).setPoint(currPoint.getX(), currPoint.getY(), -1);
				}
				((IControlPointShape) c.getShape().get()).balance();
				canvas.setTempView(ViewFactory.INSTANCE.createView(c.getShape().orElse(null)).orElse(null));
			}).
			endOrCancel((i, c) -> canvas.setTempView(null)).
			bind().activationProperty().bind(currentChoice.isEqualTo(EditionChoice.BEZIER_CURVE).and(activatedProp));
	}

	/**
	 * Binds a press interaction to add a shape.
	 */
	private void bindPressToAddShape() {
		// Add axes, grids, or dots
		nodeBinder(new Press(), AddShape.class).on(canvas).
			map(i -> {
				final IPositionShape sh = (IPositionShape) createShapeInstance();
				sh.setPosition(getAdaptedPoint(i.getSrcLocalPoint()));
				return new AddShape(sh, canvas.getDrawing());
			}).
			when(i -> i.getButton() == MouseButton.PRIMARY).
			bind().activationProperty().bind(activatedProp.and(currentChoice.isEqualTo(EditionChoice.GRID).or(currentChoice.isEqualTo(EditionChoice.DOT)).
				or(currentChoice.isEqualTo(EditionChoice.AXES))));

		// When a user starts to type a text using the text setter and then he clicks somewhere else in the canvas,
		// the text typed must be added (if possible to the canvas) before starting typing a new text.
		nodeBinder(new Press(), AddShape.class).on(canvas).
			map(i -> new AddShape(ShapeFactory.INST.createText(ShapeFactory.INST.createPoint(textSetter.getPosition()), textSetter.getTextField().getText()),
				canvas.getDrawing())).
			when(i -> textSetter.isActivated() && !textSetter.getTextField().getText().isEmpty()).
			bind().activationProperty().bind(currentChoice.isEqualTo(EditionChoice.TEXT).and(activatedProp));
	}

	/**
	 * Companion function to the bindMultiClic2AddShape binding.
	 * It initialises the first two points of the given shape.
	 */
	private IShape setInitialPtsShape(final IShape sh, final Point3D firstPt) {
		if(sh instanceof IModifiablePointsShape) {
			final IModifiablePointsShape modShape = (IModifiablePointsShape) sh;
			final IPoint pt = getAdaptedPoint(firstPt);
			modShape.setPoint(pt.getX(), pt.getY(), 0);
			modShape.setPoint(pt.getX() + 1d, pt.getY() + 1d, 1);
		}
		return sh;
	}

	/**
	 * @param shape The shape to analyse.
	 * @return The gap that must respect the pencil to not allow shape to disappear when they are too small.
	 */
	private double getGap(final IShape shape) {
		// These lines are necessary to avoid shape to disappear. It appends when the borders position is INTO.
		// In this case,the minimum radius must be computed using the thickness and the double size.
		if(shape.isBordersMovable() && shape.getBordersPosition() == BorderPos.INTO) {
			return shape.getThickness() + (shape.isDbleBorderable() && shape.hasDbleBord() ? shape.getDbleBordSep() : 0d);
		}
		return 1d;
	}

	/**
	 * Companion method to a bind. It updates the given from its centre using the new given data points.
	 */
	private void updateShapeFromCentre(final ISquaredShape shape, final IPoint startPt, final double endX) {
		final double sx = startPt.getX();
		final double radius = Math.max(sx < endX ? endX - sx : sx - endX, getGap(shape));
		shape.setPosition(sx - radius, startPt.getY() + radius);
		shape.setWidth(radius * 2d);
	}

	/**
	 * Companion method to a bind. It updates the given from its diagonal using the new given data points.
	 */
	private void updateShapeFromDiag(final IRectangularShape shape, final IPoint startPt, final IPoint endPt) {
		final double gap = getGap(shape);
		double v1 = startPt.getX();
		double v2 = endPt.getX();
		double tlx = Double.NaN;
		double tly = Double.NaN;
		double brx = Double.NaN;
		double bry = Double.NaN;
		boolean ok = true;

		if(Math.abs(v1 - v2) > gap) {
			if(v1 < v2) {
				brx = v2;
				tlx = v1;
			}else {
				brx = v1;
				tlx = v2;
			}
		}else {
			ok = false;
		}

		v1 = startPt.getY();
		v2 = endPt.getY();

		if(Math.abs(v1 - v2) > gap) {
			if(v1 < v2) {
				bry = v2;
				tly = v1;
			}else {
				bry = v1;
				tly = v2;
			}
		}else {
			ok = false;
		}

		if(ok) {
			shape.setPosition(tlx, bry);
			shape.setWidth(brx - tlx);
			shape.setHeight(bry - tly);
		}
	}

	/**
	 * @return An instance of a shape configured (thickness, colours, etc.) with the parameters of the pencil.
	 */
	public IShape createShapeInstance() {
		return setShapeParameters(currentChoice.get().createShapeInstance());
	}

	/**
	 * Configures the given shape with the parameters (e.g. thickness, colours, etc.) of the pencil.
	 * @param shape The shape to configure.
	 * @return The modified shape given as argument.
	 */
	private IShape setShapeParameters(final IShape shape) {
		shape.copy(getGroupParams());
		shape.setModified(true);
		return shape;
	}


	/**
	 * @return The current editing choice.
	 */
	public EditionChoice getCurrentChoice() {
		return currentChoice.get();
	}

	/**
	 * Sets the current editing choice.
	 * @param choice The new editing choice to set.
	 */
	public void setCurrentChoice(final EditionChoice choice) {
		if(choice != null) {
			currentChoice.set(choice);
		}
	}
}
